classes/local/store/object_file_system.php

```diff
@@ -891,4 +901,105 @@ private function exec_command($command) {
     public function copy_content_from_storedfile(stored_file $file, $target) {
         return $this->copy_file_from_hash_to_path($file->get_contenthash(), $target);
     }
+
+    /**
+     * Returns the size of the file by its contenthash.
+     *
+     * @param  string $contenthash Contenthash of the file
+     * @return mixed String or false if the file not found
+     * @throws \dml_exception
+     */
+    public function get_filesize_by_contenthash($contenthash) {
+        global $DB;
+        return $DB->get_field('files', 'filesize', ['contenthash' => $contenthash], IGNORE_MULTIPLE);
+    }
+
+    /**
+     * Gets valid HTTP ranges for range request.
+     *
+     * @param  int          $filesize Size of the file to be served.
+     * @return object|false           Array of range
+     */
+    public function get_valid_http_ranges($filesize) {
+        $range = manager::get_header($_SERVER, 'HTTP_RANGE');
+        if (!empty($range)) {
+            preg_match('{bytes=(\d+)?-(\d+)?(,)?}i', $range, $matches);
+            if (empty($matches[3])) {
+                $ranges = new \stdClass();
+                $ranges->rangefrom = (isset($matches[1])) ? intval($matches[1]) : 0;
+                $ranges->rangeto = (isset($matches[2])) ? intval($matches[2]) : ($filesize - 1);
+                $ranges->length = $ranges->rangeto - $ranges->rangefrom + 1;
+                if ($ranges->length > 5242880) {
+                    // Stream files in 5MB-chunks.
+                    $ranges->rangeto = $ranges->rangefrom + 5242880 - 1;
+                    $ranges->length = 5242880;
+                }
+                return $ranges;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Serves range request via Pre-Signed URL.
+     *
+     * @param  stored_file $file    The file to send
+     * @param  object      $ranges  Object with rangefrom, rangeto and length properties.
+     * @return false if couldn't get data.
+     * @throws \coding_exception
+     */
+    public function serve_range_request(stored_file $file, $ranges) {
+        $response = $this->curl_range_request_to_presigned_url($file->get_contenthash(), $ranges, headers_list());
+        $httpcode = manager::get_header($response['responseheaders'], 'HTTP/1.1');
+        if ($response['content'] == '' || $httpcode != '206 Partial Content') {
+            debugging('Range request failed with HTTP code: ' . $httpcode . '. Details: ' . $response['content']);
```

Comment:
can you log the failed curl url here please

Reply:
Done